src/Microsoft.ML.TimeSeries/SrCnnBatchAnomalyDetection.cs

```diff
@@ -0,0 +1,778 @@
+            }
+        }
+
+        public SrCnnBatchAnomalyDetector(IHostEnvironment env, IDataView input, string inputColumnName, string outputColumnName, double threshold, int batchSize, double sensitivity, SrCnnDetectMode detectMode)
+            : base(env, "SrCnnBatchAnomalyDetector", input)
+        {
+            Contracts.CheckValue(env, nameof(env));
+
+            Contracts.CheckValue(inputColumnName, nameof(inputColumnName));
+            _inputColumnName = inputColumnName;
+
+            env.CheckUserArg(batchSize == -1 || batchSize >= MinBatchSize, nameof(batchSize), "BatchSize must be -1 or no less than 12.");
+            if (batchSize == -1)
+            {
+                _batchSize = (int)input.GetRowCount();
+            }
+            else
+            {
+                _batchSize = batchSize;
+            }
+
+            env.CheckUserArg(threshold >= 0 && threshold <= 1, nameof(threshold), "Must be in [0,1].");
+            env.CheckUserArg(detectMode == SrCnnDetectMode.AnomalyOnly
+                || detectMode == SrCnnDetectMode.AnomalyAndExpectedValue
+                || detectMode == SrCnnDetectMode.AnomalyAndMargin, nameof(detectMode), "Invalid detectMode");
+
+            if (detectMode.Equals(SrCnnDetectMode.AnomalyOnly))
+            {
+                _outputLength = AnomalyOnlyOutputLength;
+                _modler = new SrCnnEntireModeler(threshold, sensitivity, detectMode, _outputLength);
+            }
+            else if (detectMode.Equals(SrCnnDetectMode.AnomalyAndMargin))
+            {
+                env.CheckUserArg(sensitivity >= 0 && sensitivity <= 100, nameof(sensitivity), "Must be in [0,100].");
+                _outputLength = AnomalyAndMarginOutputLength;
+                _modler = new SrCnnEntireModeler(threshold, sensitivity, detectMode, _outputLength);
+            }
+            else if (detectMode.Equals(SrCnnDetectMode.AnomalyAndExpectedValue))
+            {
+                _outputLength = AnomalyAndExpectedValueOutputLength;
+                _modler = new SrCnnEntireModeler(threshold, sensitivity, detectMode, _outputLength);
+            }
+
+            SchemaBindings = new Bindings(input.Schema, inputColumnName, outputColumnName, new VectorDataViewType(NumberDataViewType.Double, _outputLength));
+        }
+
+        protected override ColumnBindingsBase SchemaBindings { get; }
+
+        protected override Delegate[] CreateGetters(DataViewRowCursor input, Batch currentBatch, bool[] active)
+        {
+            if (!SchemaBindings.AnyNewColumnsActive(x => active[x]))
+                return new Delegate[1];
+            return new[] { currentBatch.CreateGetter(input, _inputColumnName) };
+        }
+
+        protected override Batch InitializeBatch(DataViewRowCursor input) => new Batch(_batchSize, _outputLength, _modler);
+
+        protected override Func<bool> GetIsNewBatchDelegate(DataViewRowCursor input)
```

Comment:
>GetIsNewBatchDelegate [](start = 38, length = 21)

I think this method and `GetLastInBatchDelegate` can be implemented in the base class, I can't imagine a future implementation of the base class that would use a different way of computing this. @harishsk, what do you think?

Reply:
Since a check for _batchSize==-1 is added here, I think it's proper to keep it in inherit class

---
In reply to: [426409875](https://github.com/dotnet/machinelearning/pull/5135#discussion_r426409875) [](ancestors = 426409875)

Reply:
I think adding -1 as an option meaning "the entire data view" is reasonable for the base class, but I won't insist on it if you think that it is something specific to this implementation of the base class.
/cc @harishsk 

---
In reply to: [427135642](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427135642) [](ancestors = 427135642,426409875)

Reply:
Using -1 as the "entire data view" is reasonable. But if @mengaims feels it is proper to keep it in the inherited class, I am okay with it since we only have one derived class using it now. If we get another inherited class, we can revisit this then and refactor if necessary. Since this is an internal implementation detail, there shouldn't be any issues.

---
In reply to: [427278034](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427278034) [](ancestors = 427278034,427135642,426409875)